### DISCUSSION TITLE
Confusion from the readme

### DISCUSSION THREAD

**Discussion by chazste** [PlummersSoftwareLLC/HelloAssembly#21] [Q&A] [answered]
The size table at the end of the readme does not include the MASM32 Link size of TinyOriginal and shows that its size with crinkler is smaller than the shell code version.

I see some opportunity to improve still but it is easier to think about the pre-crinkler version and sizes.

Thanks!

**Comment by rbergen** [comment, marked as answer]
Your observations about the size table are correct:

- The latest version(s) of the TinyOriginal code fail to link with the MASM32 linker, which is why that combination is no longer included in the table. (This does indeed mean that "TinyOriginal" is no longer entirely "original", in the sense that it has been size-optimized itself.)
- The Crinkled version of the TinyOriginal executable is indeed smaller than the shell code version; it's true that the video can be interpreted otherwise. We have a loose theory about the cause of this, which is that the shell code approach confuses Crinkler such that some of its optimizations don't work/apply.

If you want to think about pre-Crinkler versions from a code size perspective, then I think it is now best to look at obj file sizes, so after assembly but before linking. From the perspective of the challenge this repo is connected to, the goal is the smallest executable. That is clearly achieved with Crinkler being used with the (optimized) TinyOriginal code.

As your observations are correct but don't translate to an error in the readme (or code), I've closed this as an issue and converted it to a Q&A topic in Discussions.

**Comment by chazste** [comment]
Cool thanks!

I'll have to recheck. I was able to link TinyOriginal, but I don't always use the linker from MASM32. (I usually use the linker from VS2022)

Thanks again for the hints. Hopefully I can submit something useful soon.

**Reply by rbergen** [reply]
You're welcome. I should have said the MASM32 linking fails on at least one of my machines. Your mileage may indeed vary.

**Reply by chazste** [reply]
I have done some testing using the Microsoft linker both from masm32 and VS2022. I have a branch with a batch file that works for me and makes linking with crinkler optional. I can open a PR if this is desirable; but even if it isn't I would appreciate it if you could give the batch file a try and see if the masm32 link program works for you with that command line. I have tested it on my computer and in the MS sandbox application with just masm32 SDK.

**Reply by rbergen** [reply]
If you paste the masm32 command line here I'll give it a try when I have the time. It if works for me I can reintroduce masm32 linking in the build scripts and the size table.

**Comment by chazste** [comment]
link /LIBPATH:"C:\masm32\lib" /NODEFAULTLIB:LIBCMT.lib /ENTRY:MainEntry /DEBUG:NONE /subsystem:windows /map /merge:.rdata=.text /merge:.data=.text /align:16 gdi32.lib user32.lib kernel32.lib tiny.obj

The result runs and is:
01/28/2023  01:44 PM             1,456 tiny.exe

I see it slightly larger (32 bytes) if I link against the libraries in the Windows 10 SDK.

**Reply by rbergen** [reply]
When I use the command line you shared with the MASM32 linker (c:\masm32\bin\link.exe) I do get an executable of the same size (1456 bytes), but the application doesn't show when I try to run it.
